pennylane/vqe/vqe.py

```diff
@@ -420,20 +423,58 @@ def ansatz(params, **kwargs):
     """
 
     def __init__(
-        self, ansatz, hamiltonian, device, interface="autograd", diff_method="best", **kwargs
+        self,
+        ansatz,
+        hamiltonian,
+        device,
+        interface="autograd",
+        diff_method="best",
+        optimize=False,
+        **kwargs,
     ):
         coeffs, observables = hamiltonian.terms
+
         self.hamiltonian = hamiltonian
         """Hamiltonian: the hamiltonian defining the VQE problem."""
 
-        self.qnodes = qml.map(
-            ansatz, observables, device, interface=interface, diff_method=diff_method, **kwargs
-        )
-        """QNodeCollection: The QNodes to be evaluated. Each QNode corresponds to the
-        the expectation value of each observable term after applying the circuit ansatz.
-        """
+        self.qnodes = None
+        """QNodeCollection: The QNodes to be evaluated. Each QNode corresponds to the the 
+        expectation value of each observable term after applying the circuit ansatz."""
+
+        if optimize:
+            if not qml.tape_mode_active():
+                raise ValueError(
+                    "Observable optimization is only supported in tape mode. Tape "
+                    "mode can be enabled with the command:\n"
+                    "qml.enable_tape()"
+                )
+
+            obs_groupings, coeffs_groupings = qml.grouping.group_observables(observables, coeffs)
+
+            wires = device.wires.tolist()
+
+            @qml.qnode(device, interface=interface, diff_method=diff_method, **kwargs)
+            def circuit(*qnode_args, obs, **qnode_kwargs):
+                """Converting ansatz into a full circuit including measurements"""
+                ansatz(*qnode_args, wires=wires, **qnode_kwargs)
+                return [qml.expval(o) for o in obs]
+
+            def cost_fn(*qnode_args, **qnode_kwargs):
+                """Combine results from grouped QNode executions with grouped coefficients"""
+                total = 0
+                for o, c in zip(obs_groupings, coeffs_groupings):
+                    res = circuit(*qnode_args, obs=o, **qnode_kwargs)
+                    total += sum([r * c_ for r, c_ in zip(res, c)])
+                return total
+
+            self.cost_fn = cost_fn
+
+        else:
+            self.qnodes = qml.map(
+                ansatz, observables, device, interface=interface, diff_method=diff_method, **kwargs
+            )
 
-        self.cost_fn = qml.dot(coeffs, self.qnodes)
+            self.cost_fn = qml.dot(coeffs, self.qnodes)
 
     def __call__(self, *args, **kwargs):
         return self.cost_fn(*args, **kwargs)
```

Comment:
@josh146, I may need some help on the metric tensor part.

Reply:
@trbromley Do you mean the `metric_tensor` method? 🤔 

In the short term, could we just have 

```python
def metric_tensor(self, args, kwargs=None, diag_approx=False, only_construct=False):
    if self.qnodes is None:
        raise SomeSpecificError("The metric tensor cannot be calculated when optimize=True")
```

Reply:
Yes! The current method just gives the metric tensor of the first QNode. Could we also do that in the `optimize=True` approach? In that case, the QNodes may return multiple values of varying length, so I'm not sure if that's a problem.

Reply:
> Could we also do that in the optimize=True approach?

The issue is the metric tensor computation would need to be rewritten, since it likely will fail and/or return the incorrect result, because it was not written assuming overlapping observables (could be wrong though).

Since the metric tensor is due to be changed into a transform, might be best to just wait for now, and go with the short term solution (no metric tensor when `optimize=true`)

Reply:
Thanks! Ok I just prevented access to the metric tensor in `optimize=True` mode for now.